listen/array-changes.js

```diff
@@ -254,12 +267,33 @@ var observableArrayProperties = {
 
     push: {
         value: function push(arg) {
-            if (arguments.length === 1) {
-                return this.splice(this.length, 0, arg);
-            } else {
-                var args = array_slice.call(arguments);
-                return this.swap(this.length, 0, args);
-            }
+			var start = this.length, 
+				addedCount = arguments.length, 
+				argArray,
+				hasOwnPropertyChangeDescriptor;
+			
+           	argArray = addedCount === 1 ? [arguments[0]] : Array.apply(null, arguments);
+           	//argArray = array_slice.call(arguments);
+
+			if(addedCount > 0) {
+	    		PropertyChanges.dispatchBeforeOwnPropertyChange(this, "length", start);
+				this.dispatchBeforeRangeChange(argArray, Array.empty, start);
+				
+				if(hasOwnPropertyChangeDescriptor = PropertyChanges.hasOwnPropertyChangeDescriptor(this)) {
+					this._dispatchBeforeOwnPropertyChange(start, addedCount);
+				}
+			}
+			
+			array_push.apply(this,arguments);
+
+	        if (addedCount > 0) {
+				if (hasOwnPropertyChangeDescriptor) {
+					this._dispatchOwnPropertyChange(start,addedCount);
+				}
+		        this.dispatchRangeChange(argArray,Array.empty, start);
+				this.dispatchOwnPropertyChange("length", this.length);
+	        }
+
```

Comment:
This is a reasonable case to unroll. In the v2 branch, I’m toying with changing the signature of swap to be the better kernel for all mutations, as documented here http://github.com/kriskowal/pop-swap